### PR TITLE
Refactor Scribe

### PR DESCRIPTION
**Pull request by nautsimon**
**Description**
- [x] Refactor filter.go: Reduced channel hell. Took out excessive default case, provided more race condition protection. More intuitive flow with batching and get logs.
- [x] Refactor config: more intuitive, more reorg protection options
- [x] Add Otel monitoring (https://github.com/synapsecns/sanguine/pull/1086), (https://github.com/synapsecns/sanguine/issues/874) (high prio)
- [x] Add more tests/break down big functions
- [x] Livefill refactor w/more tests
- [x] A livefill test that uses a prod level config

---
- [x] Make sure still works with agents (https://github.com/synapsecns/sanguine/pull/1086)

---
stretch goals (for likely other PRs)
1.  redo logging and error handling: error codes, less confusing logs
2. redo/update readme
3. provide option to categorize contracts in config (for logging)
4. clean up testing suites
5. clean up/beef up comments
6. enable concurrent livefill and backfill. This would involve splitting up config with a "livefill" and "backfill" section. This way ranges can be re-backfilled/backfilled while continuing to livefill

### CODE REVIEW

File: services/scribe/backfill/contract.go [hunk 1]
```diff
@@ -103,25 +104,27 @@ func (c *ContractBackfiller) Backfill(parentCtx context.Context, givenStart uint
...
 			select {
 			case <-groupCtx.Done():
 				LogEvent(ErrorLevel, "Context canceled while storing and retrieving logs", LogData{"cid": c.chainConfig.ChainID, "ca": c.contractConfig.Address})
 
 				return fmt.Errorf("context canceled while storing and retrieving logs: %w", groupCtx.Err())
-			case log := <-logsChan:
+			case log, ok := <-logsChan:
+				if !ok {
+					return nil
+				}
```
Comment (nautsimon) [inline services/scribe/backfill/contract.go:127]: No more doneChan, instead there is a listener for the channel being closed and empty. This goroutine is exited when this condition is met.

There is another channel but specifically for catching errors for better logging. The previous channel structure was susceptible to race conditions where a value from the doneChan could be received before all logs had been removed from the logsChan.

File: services/scribe/backfill/contract.go [hunk 2]
```diff
@@ -339,30 +338,28 @@ func (c *ContractBackfiller) getLogs(parentCtx context.Context, startHeight, end
...
+			case <-logFetcher.doneChan:
+				remLogChunks, err := logFetcher.Drain(ctx)
+				if err != nil {
+					LogEvent(ErrorLevel, "Error draining logs", LogData{"cid": c.chainConfig.ChainID, "sh": startHeight, "eh": endHeight, "e": err.Error()})
 					return
 				}
+				for _, log := range remLogChunks {
+					logsChan <- log
+				}
+				close(logsChan)
```
Comment (nautsimon) [inline services/scribe/backfill/contract.go:356]: Closing the logsChan, once the channel is empty, the consuming goroutine at the top of this file exits its loop.

File: services/scribe/backfill/fetcher.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
+package backfill
```
Comment (nautsimon) [inline services/scribe/backfill/fetcher.go:1]: Renamed this file to fetcher as it seemed more appropriate.

File: services/scribe/backfill/fetcher.go [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+// Start starts the log fetching process. If the context is canceled, logs will stop being filtered.
+// 1. Within an infinite for loop, chunks of getLogs blocks are constructed and used to get logs. This flow is paused
+// when the logs channel's buffer of 15 is reached.
+// 2. Each time the logs are received, a wait group is used to ensure that there is no race condition
+// where channels could be closed before a log could be saved.
+// 3. When the range to get logs is completed (GetChunkArr returns a zero array), the wait group is used to ensure
+// that all logs are added to the logs channel before returning and terminating the function.
+// 4. Completing the Start function triggers the closeOnDone function, which sends a boolean in the done channel
+// that signals that the fetcher has completed. The consumer of these logs then performs a drain to fully empty the logs
+// channel. See contract.go to learn more how the logs from this file are consumed.
```
Comment (nautsimon) [inline services/scribe/backfill/fetcher.go:99]: explanation of the new flow

File: services/scribe/backfill/fetcher.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+		}
+		chunkArr = append(chunkArr, chunk)
+
+		// Stop appending chunks if the max height of the current chunk exceeds the concurrency threshold
+		if chunk.EndBlock.Uint64() > f.endBlock.Uint64()-f.chainConfig.ConcurrencyThreshold {
+			return chunkArr
+		}
+	}
+	return chunkArr
+}
```
Comment (nautsimon) [inline services/scribe/backfill/fetcher.go:88]: Previously, a chunk size and sub chunk size was specified in the config. This was convoluted and confusing.

Now you specify the amount of get logs requests you want to batch and the range of each. This function creates an array of block ranges for each getlogs request (using an iterator). The batching backend was slightly refactored to take an array of chunks instead of making another iterator.

File: services/scribe/backfill/fetcher_test.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
+package backfill_test
```
Comment (nautsimon) [inline services/scribe/backfill/fetcher_test.go:1]: More tests so we know whats going on

File: services/scribe/backfill/backend.go [hunk 7]
```diff
@@ -100,3 +94,38 @@ func GetLogsInRange(ctx context.Context, backend ScribeBackend, startHeight uint
 
 	return res.List(), nil
 }
+
+// BlockHashesInRange gets all block hashes in a range with a single batch request
+// in successful cases an immutable map is returned of [height->hash], otherwise an error is returned.
+func BlockHashesInRange(ctx context.Context, backend ScribeBackend, startHeight uint64, endHeight uint64) (*immutable.Map[uint64, string], error) {
```
Comment (nautsimon) [inline services/scribe/backfill/backend.go:100]: For reducing requests for doing reorg checks (batching)

File: services/scribe/node/scribe.go [hunk 8]
```diff
@@ -44,28 +45,30 @@ func NewScribe(eventDB db.EventDB, clients map[uint32][]backfill.ScribeBackend,
...
 		chainID := chainConfig.ChainID
 
+		// Set default confirmation values
+		if chainConfig.ConfirmationConfig.RequiredConfirmations == 0 ||
+			chainConfig.ConfirmationConfig.ConfirmationThreshold == 0 ||
+			chainConfig.ConfirmationConfig.ConfirmationRefreshRate == 0 {
+			chainConfig.ConfirmationConfig = config.ConfirmationConfig{
+				RequiredConfirmations:   250,
+				ConfirmationThreshold:   100,
+				ConfirmationRefreshRate: 5,
```
Comment (nautsimon) [inline services/scribe/node/scribe.go:67]: Default config for reorg protection

File: services/scribe/node/scribe.go [hunk 9]
```diff
@@ -116,192 +116,173 @@ func (s Scribe) Start(ctx context.Context) error {
...
-// TODO: Has issues with last confirmed data. Needs to be fixed.
+// confirmBlocks checks for reorgs with data stored in the scribe database
+// 0. Every few seconds (depending on ConfirmationRefreshRate in the config), the confirmBlocks function is called.
+// 1. First, the head block and the last confirmed block is retrieved. There is no "backfill" capability for reorgs.
+// it is suggested you reindex that range you want to confirm if it is far from the head.
+// 2. Block hashes for the blocks since the last confirmed block up to (latest block - the confirmation threshold)
+// are batch requested.
+// 3. These hashes are used to query the databases for receipts that do not have those block hashes.
+// 4. The returned receipts have their blocks deleted and re-backfilled.
+// 5. The entire range of blocks is then confirmed and last confirmed is updated.
```
Comment (nautsimon) [inline services/scribe/node/scribe.go:127]: How the new livefill reorg protection works

File: services/scribe/node/scribe.go [hunk 10]
```diff
@@ -116,192 +116,173 @@ func (s Scribe) Start(ctx context.Context) error {
...
+func (s Scribe) confirmBlocks(ctx context.Context, chainConfig config.ChainConfig) error {
+	chainID := chainConfig.ChainID
+	requiredConfirmations := chainConfig.ConfirmationConfig.RequiredConfirmations
+	getBlockBatchAmount := chainConfig.GetBlockBatchAmount
+	if getBlockBatchAmount == 0 {
+		getBlockBatchAmount = 25
+	}
+	confirmationThreshold := chainConfig.ConfirmationConfig.ConfirmationThreshold
+
+	latestBlock, err := s.clients[chainID][0].BlockNumber(ctx)
```
Comment (nautsimon) [inline services/scribe/node/scribe.go:139]: Essentially, blocks between the last confirmed block (1000 below the head if no last confirmed block) and latest block - the confirmation threshold will be checked.

File: services/scribe/node/scribe.go [hunk 11]
```diff
@@ -116,192 +116,173 @@ func (s Scribe) Start(ctx context.Context) error {
...
-		if err != nil {
-			logger.Errorf("[LIVEFILL] could not retrieve receipts with filter %d chain: %d, block: %d, %v", newBlock-uint64(requiredConfirmations), chainID, i, err)
+	confirmTo := latestBlock - confirmationThreshold
+	confirmFrom := lastConfirmedBlock + 1
+	blockHashes, err := GetBlockHashes(ctx, s.clients[chainID][0], confirmFrom, confirmTo, getBlockBatchAmount)
+	if err != nil {
+		return fmt.Errorf("could not get blockHashes: %w", err)
+	}
+	// get receipts emitted on invalid block hashes
+	invalidReceipts, err := s.eventDB.RetrieveReceiptsWithStaleBlockHash(ctx, chainID, blockHashes, confirmFrom, confirmTo)
```
Comment (nautsimon) [inline services/scribe/node/scribe.go:167]: Does all parity checks in SQL now

File: services/scribe/node/scribe_test.go [hunk 12]
```diff
@@ -15,13 +19,114 @@ import (
...
+		latestBlock, err := backend.BlockNumber(ctx)
+		Nil(l.T(), err)
+
+		if latestBlock >= desiredBlockHeight {
+			return address
+		}
+	}
+}
+func (l *LiveSuite) TestGetBlockHashes() {
+	testBackend := geth.NewEmbeddedBackend(l.GetTestContext(), l.T())
```
Comment (nautsimon) [inline services/scribe/node/scribe_test.go:88]: Tests 
- blockhash batching
- livefill for 20 sec and checks db for content
- livefill 20 sec and checks db for content, verifies confirmations, checks latest block stats
- livefill for 20 sec with a receipt with an invalid blockhash, verifies it gets removed
- adding a livefill test with a prod config

**Review by nautsimon — COMMENTED** [review]
key items so far

File: services/scribe/backfill/contract_test.go [hunk 13]
```diff
@@ -248,86 +249,88 @@ func (b BackfillSuite) TestContractBackfill() {
...
-	logs, err := b.testDB.RetrieveLogsWithFilter(b.GetTestContext(), db.LogFilter{}, 1)
-	Nil(b.T(), err)
-	Equal(b.T(), 4, len(logs))
-	receipts, err := b.testDB.RetrieveReceiptsWithFilter(b.GetTestContext(), db.ReceiptFilter{}, 1)
-	Nil(b.T(), err)
-	Equal(b.T(), 1, len(receipts))
-}
-
-// TestTxTypeNotSupported tests how the contract backfiller handles a transaction type that is not supported.
+//	func (b BackfillSuite) TestTxTypeNotSupported() {
```
Comment (trajan0x) [inline services/scribe/backfill/contract_test.go:256]: We shouldn't be adding less tests here. If anything, a lot of this codebase is lacking tests around edge cases 

Maybe we can fetch these txes as json from the rpcs and embed them as constants in our tests so we don't need to do the network fetches?

File: services/scribe/backfill/contract_test.go [hunk 14]
```diff
@@ -248,86 +249,88 @@ func (b BackfillSuite) TestContractBackfill() {
...
+//		receipts, err := b.testDB.RetrieveReceiptsWithFilter(b.GetTestContext(), db.ReceiptFilter{}, 1)
+//		Nil(b.T(), err)
+//		Equal(b.T(), 1, len(receipts))
+//	}
+//
+// // TestTxTypeNotSupported tests how the contract backfiller handles a transaction type that is not supported.
+// //
+// // nolint:dupl
+//
+//	func (b BackfillSuite) TestInvalidTxVRS() {
```
Comment (trajan0x) [inline services/scribe/backfill/contract_test.go:297]: see above

File: services/scribe/backfill/fetcher.go (hunk 5 above)
Comment (trajan0x) [reply services/scribe/backfill/fetcher.go:88]: the real fix to the concurrency limits has to come in server.go. Namely, we need to make sure logs we get are less than the lastIndexed for the given contract

File: services/scribe/backfill/fetcher.go [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+			}
+			logs, err := f.FilterLogs(ctx, chunks)
+			if err != nil {
+				return fmt.Errorf("could not filter logs: %w", err)
+			}
+
+			wg.Add(1)
+			go func(logs []types.Log) {
+				defer wg.Done()
+				f.logs <- logs
```
Comment (trajan0x) [inline services/scribe/backfill/fetcher.go:129]: will leak memory if context not cancelled, use select

File: services/scribe/backfill/fetcher.go [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+			}(logs)
+			LogEvent(InfoLevel, "Contract backfill chunk completed", LogData{"ca": f.contractAddress, "sh": chunks[0].MinBlock(), "eh": chunks[0].MaxBlock(), "cid": &f.chainConfig.ChainID})
```
Comment (trajan0x) [inline services/scribe/backfill/fetcher.go:131]: wtf are ca, sh, eh, etc

File: services/scribe/backfill/fetcher.go [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+				}
+			}
+
+			return logs, nil
+		}
+	}
+}
+
+// Drain fetches empties the log chan. For use once the doneChan is emitted.
+func (f *RangeFilter) Drain(ctx context.Context) (filteredLogs []types.Log, err error) {
```
Comment (trajan0x) [inline services/scribe/backfill/fetcher.go:199]: does this need to be called by external users? If so why? If not, let's make it private

File: services/scribe/backfill/fetcher_test.go [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+		StartBlock: big.NewInt(1),
+		EndBlock:   big.NewInt(10),
+	}}
+	logInfo, err := rangeFilter.FilterLogs(b.GetTestContext(), chunks)
+	Nil(b.T(), logInfo)
+	NotNil(b.T(), err)
+}
+
+// TestGetChunkArr ensures that the batching orchestration function (collecting block range chunks into arrays) works properly.
+func (b BackfillSuite) TestGetChunkArr() {
```
Comment (trajan0x) [inline services/scribe/backfill/fetcher_test.go:48]: Can we emulate a race condition on donechan as a test?

File: services/scribe/cmd/cmd.md [hunk 19]
```diff
@@ -20,6 +20,18 @@ $ backfill --config </Full/Path/To/Config.yaml> --db <sqlite or mysql> --path <p
...
 ```
 
+## Configuration
+chain_id: The ID of the chain
+required_confirmations: the number of confirmations required for a block to be finalized
+contracts: stores all the contract information for the chain.
+get_logs_range: is the number of blocks to request in a single getLogs request.
+get_logs_batch_amount: is the number of getLogs requests to include in a single batch request.
+store_concurrency: is the number of goroutines to use when storing data.
+concurrency_threshold: is the max number of block from head in which concurrent operations (store, getlogs) is allowed.
```
Comment (trajan0x) [inline services/scribe/cmd/cmd.md:30]: not at all clear this is used in yaml

File: services/scribe/cmd/cmd.md [hunk 20]
```diff
@@ -20,6 +20,18 @@ $ backfill --config </Full/Path/To/Config.yaml> --db <sqlite or mysql> --path <p
 $ server --port <port> --db <sqlite or mysql> --path <path/to/database or database url>
 ```
 
+## Configuration
```
Comment (trajan0x) [inline services/scribe/cmd/cmd.md:23]: space

File: services/scribe/config/chain.go [hunk 21]
```diff
@@ -7,26 +7,36 @@ import (
...
 	Contracts ContractConfigs `yaml:"contracts"`
-	// BlockTimeChunkCount is the number of chunks (goroutines) to process at a time while backfilling blocktimes.
-	BlockTimeChunkCount uint64 `yaml:"block_time_chunk_count"`
-	// BlockTimeChunkSize is the number of blocks to process per chunk (goroutine) while backfilling blocktimes.
-	BlockTimeChunkSize uint64 `yaml:"block_time_chunk_size"`
-	// ContractSubChunkSize is the number of blocks to request for in each get logs request in the batch request.
-	ContractSubChunkSize int `yaml:"contract_sub_chunk_size"`
-	// ContractChunkSize is the number of blocks to process per chunk while backfilling contracts.
-	ContractChunkSize int `yaml:"contract_chunk_size"`
+	// GetLogsRange is the number of blocks to request in a single getLogs request.
```
Comment (trajan0x) [inline services/scribe/config/chain.go:28]: max number*

File: services/scribe/db/datastore/sql/base/receipt.go [hunk 22]
```diff
@@ -238,3 +237,18 @@
 
 	return count, nil
 }
+
+// RetrieveReceiptsWithStaleBlockHash gets receipts that are from a reorged/stale block.
+func (s Store) RetrieveReceiptsWithStaleBlockHash(ctx context.Context, chainID uint32, blockHashes []string, startBlock uint64, endBlock uint64) ([]types.Receipt, error) {
+	var dbReceipts []Receipt
+	dbTx := s.DB().WithContext(ctx).Model(&Receipt{}).Where("block_number >= ? ", startBlock).Where("block_number <= ? ", endBlock).Where("block_hash NOT IN (?)", blockHashes).Scan(&dbReceipts)
```
Comment (trajan0x) [inline services/scribe/db/datastore/sql/base/receipt.go:244]: use the namer for this

Comment (trajan0x) [inline services/scribe/db/datastore/sql/base/receipt.go:244]: use the namer for this, also enter between .Where's will be more readable imo

File: services/scribe/db/receipt_test.go [hunk 23]
```diff
@@ -186,3 +187,33 @@ func (t *DBSuite) MakeRandomReceipt(txHash common.Hash) types.Receipt {
...
+		hash2 := common.BigToHash(big.NewInt(2))
+		hash3 := common.BigToHash(big.NewInt(3))
+		newHash2 := common.BigToHash(big.NewInt(22))
+		blockHashes := []common.Hash{hash1, hash2, hash3}
+		blockHashesStr := []string{hash1.String(), newHash2.String(), hash3.String()}
+
+		for i := 0; i < 10; i++ {
+			receipt := t.MakeRandomReceipt(common.BigToHash(big.NewInt(gofakeit.Int64())))
+			receipt.BlockNumber = big.NewInt(int64(i))
+			fmt.Println("SSS", i, blockHashes[i%3])
```
Comment (trajan0x) [inline services/scribe/db/receipt_test.go:205]: ??

File: services/scribe/db/suite_test.go [hunk 24]
```diff
@@ -60,15 +60,20 @@ func (t *DBSuite) SetupSuite() {
...
+func (t *DBSuite) connString(dbname string) string {
+	return fmt.Sprintf("%s:%s@tcp(%s:%d)/%s?parseTime=true", core.GetEnv("MYSQL_USER", "root"), os.Getenv("MYSQL_PASSWORD"), core.GetEnv("MYSQL_HOST", "127.0.0.1"), core.GetEnvInt("MYSQL_PORT", 3306), dbname)
+}
+
 func (t *DBSuite) setupMysqlDB() {
 	// skip if mysql test disabled, this really only needs to be run in ci
 
 	// skip if mysql test disabled
-	if os.Getenv(dbcommon.EnableMysqlTestVar) == "" {
+	if os.Getenv("ENABLE_MYSQL_TEST") == "" {
```
Comment (trajan0x) [inline services/scribe/db/suite_test.go:72]: why?

File: services/scribe/db/suite_test.go [hunk 25]
```diff
@@ -60,15 +60,20 @@ func (t *DBSuite) SetupSuite() {
...
 		return
 	}
 	// sets up the conn string to the default database
-	connString := dbcommon.GetTestConnString()
+	connString := t.connString(os.Getenv("MYSQL_DATABASE"))
```
Comment (trajan0x) [inline services/scribe/db/suite_test.go:76]: see above

File: services/scribe/node/export_test.go [hunk 26]
```diff
@@ -1,8 +1,6 @@
 package node
 
-import "context"
-
-// ProcessRange exports filtering logs for testing.
-func (s Scribe) ProcessRange(ctx context.Context, chainID uint32, requiredConfirmations uint32) error {
-	return s.confirmBlocks(ctx, chainID, requiredConfirmations)
-}
+//// ProcessRange exports filtering logs for testing.
```
Comment (trajan0x) [inline services/scribe/node/export_test.go:3]: if nothing needs to be exported for scribe/node tests just delete the file

File: services/scribe/node/scribe_test.go [hunk 27]
```diff
@@ -15,13 +19,114 @@ import (
...
 	"math/big"
 	"os"
+	"sync"
 	"time"
 )
 
+// TODO combine these functions with backfill/backend as well as other tests
+
+// ReachBlockHeight reaches a block height on a backend.
+func (l *LiveSuite) ReachBlockHeight(ctx context.Context, backend backends.SimulatedTestBackend, desiredBlockHeight uint64) {
```
Comment (trajan0x) [inline services/scribe/node/scribe_test.go:29]: could also put this in ethergo/backends/base maybe?

File: services/scribe/node/scribe_test.go [hunk 28]
```diff
@@ -152,67 +263,369 @@ func (l LiveSuite) TestRequiredConfirmationSetting() {
...
+	Nil(l.T(), err)
+	Equal(l.T(), 9-chainConfig.ConfirmationConfig.ConfirmationThreshold, lastConfirmedBlock)
+
+	lastBlockIndexed, err := l.testDB.RetrieveLastIndexed(l.GetTestContext(), testContract.Address(), chainConfig.ChainID)
+	Nil(l.T(), err)
+	Equal(l.T(), uint64(9), lastBlockIndexed)
 }
+
+// TODO finish this test
+// TestLivefillParity runs livefill on certain prod chains. Then it checks parity with an explorer API.
```
Comment (trajan0x) [inline services/scribe/node/scribe_test.go:410]: si

File: services/scribe/testhelper/scribe.go [hunk 29]
```diff
@@ -59,9 +59,8 @@ func NewTestScribe(ctx context.Context, tb testing.TB, deployedContracts map[uin
 
 		// add the chain config to the list
 		chainConfigs = append(chainConfigs, config.ChainConfig{
-			ChainID:               uint32(backend.GetChainID()),
-			RequiredConfirmations: 1,
-			Contracts:             contractConfigs,
+			ChainID:   uint32(backend.GetChainID()),
+			Contracts: contractConfigs,
```
Comment (trajan0x) [inline services/scribe/testhelper/scribe.go:63]: don't we still want required confirmations in the config for the server responses?

**Review by trajan0x — CHANGES_REQUESTED** [review]
See comments

File: services/scribe/backfill/backend.go [hunk 30]
```diff
@@ -12,16 +12,15 @@
...
 	"math/big"
 )
 
 // ScribeBackend is the set of functions that the scribe needs from a client.
 type ScribeBackend interface {
 	// ChainID gets the chain id from the rpc server.
 	ChainID(ctx context.Context) (*big.Int, error)
-	// BlockByNumber retrieves a block from the database by number, caching it
-	// (associated with its hash) if found.
+	// BlockByNumber retrieves a block
```
[the remark this answers is not in the document]
Comment (nautsimon) [reply services/scribe/backfill/backend.go:23]: Im unsure who karalabe is, this is an artifact of whoever created the ScribeBackend interface.

File: services/scribe/backfill/chain.go [hunk 31]
```diff
@@ -67,8 +68,8 @@
 		refreshRate = 1
 	}
 
-	if chainConfig.StoreConcurrencyThreshold == 0 {
-		chainConfig.StoreConcurrencyThreshold = 500
+	if chainConfig.ConcurrencyThreshold == 0 {
+		chainConfig.ConcurrencyThreshold = 5000
 	}
 	minBlockHeight := uint64(math.MaxUint64)
```
[the remark this answers is not in the document]
Comment (nautsimon) [reply services/scribe/backfill/chain.go:74]: the concurrency threshold value is being used for not just stopping concurrent stores near the head but also batching get logs now, so I removed the "store" from this value

the bump in the default value is to account for chains that might be doing massive getlogs ranges.

File: services/scribe/backfill/chain.go [hunk 32]
```diff
@@ -144,7 +145,9 @@
 							return fmt.Errorf("could not get current block number while backfilling: %w", err)
 						}
 					}
-
+					if onlyOneBlock != nil {
```
[the remark this answers is not in the document]
Comment (nautsimon) [reply services/scribe/backfill/chain.go:148]: this was for testing.

see dm

File: services/scribe/backfill/contract.go [hunk 33]
```diff
@@ -339,30 +338,28 @@
...
-					LogEvent(ErrorLevel, "range filter drained", LogData{"cid": c.chainConfig.ChainID, "ca": c.contractConfig.Address, "len": len(finLogs), "err": err})
-
-					for _, log := range finLogs {
-						LogEvent(ErrorLevel, "Log filter done, adding log to chan", LogData{"cid": c.chainConfig.ChainID, "bn": log.BlockNumber, "tx": log.TxHash.Hex(), "la": log.Address.String(), "ca": c.contractConfig.Address})
-
-						logsChan <- log
-					}
-					doneChan <- true
+			case <-logFetcher.doneChan:
+				remLogChunks, err := logFetcher.Drain(ctx)
```
[the remark this answers is not in the document]
Comment (nautsimon) [reply services/scribe/backfill/contract.go:348]: yeah with the check in fetcher, drain is obsolete

File: services/scribe/backfill/contract_test.go (hunk 13 above)
Comment (nautsimon) [reply services/scribe/backfill/contract_test.go:256]: was commented out so i could run new tests faster locally, will un comment pre-merge

File: services/scribe/backfill/fetcher.go (hunk 16 above)
Comment (nautsimon) [reply services/scribe/backfill/fetcher.go:131]: its for mapping attributes for logging (see logger)

obviously, this is overboard, its in the description of this pr to get do this with error codes (as a stretch goal)

File: services/scribe/db/receipt_test.go (hunk 23 above)
Comment (nautsimon) [reply services/scribe/db/receipt_test.go:205]: is this questioning the print or something else

Comment (nautsimon) [reply services/scribe/db/receipt_test.go:205]: is this questioning the printout or something else

File: services/scribe/testhelper/scribe.go (hunk 29 above)
Comment (nautsimon) [reply services/scribe/testhelper/scribe.go:63]: for omnirpc? yeah that makes sense

**Review by CryptoMaxPlanck — APPROVED** [review]
Simon says
